### PR TITLE
US015 & US012 & US020 add, edit, & delete role

### PR DESCRIPTION
**Pull request by joshuasy10**

### CODE REVIEW

File: src/main/java/org/kainos/ea/exception/BandDoesNotExistException.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package org.kainos.ea.exception;
+
+public class BandDoesNotExistException extends Throwable {
+    @Override
+    public String getMessage(){
+        return "Capability does not exist";
```
Comment (jedgar3) [inline src/main/java/org/kainos/ea/exception/BandDoesNotExistException.java:6]: This should return band does not exist rather than capability

**Review by jedgar3 — CHANGES_REQUESTED** [review]

File: src/main/java/org/kainos/ea/exception/BandDoesNotExistException.java (hunk 1 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/exception/BandDoesNotExistException.java]: forgot to push that change

File: src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 2]
```diff
@@ -47,4 +46,52 @@ public Response getJobRoleById(@PathParam("id") int id) {
...
-}
+
+    @POST
+    @Path("/job-roles")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response createJobRole(JobRoleRequest jobRoleRequest) {
+        try {
+            jobRoleService.createJobRole(jobRoleRequest);
+            return Response.status(Response.Status.CREATED).build();
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/controller/JobRoleController.java:57]: you probably want to return and id from this endpoint so people know how to find the role they have created

File: src/main/java/org/kainos/ea/dao/BandDao.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class BandDao {
+    public List<Band> getAllBands() throws SQLException, DatabaseConnectionException {
+        Connection c = DatabaseConnector.getConnection();
+
+        String selectStatement = "SELECT * FROM bands;";
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/dao/BandDao.java:16]: no wildcards

File: src/main/java/org/kainos/ea/service/CapabilityService.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package org.kainos.ea.service;
+
+import org.kainos.ea.dao.CapabilityDao;
+import org.kainos.ea.exception.*;
+import org.kainos.ea.model.Capability;
+import java.sql.SQLException;
+import java.util.List;
+
+public class CapabilityService {
+    private final CapabilityDao jobRoleDao;
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/service/CapabilityService.java:10]: bad var name

File: src/test/java/integration/controller/JobRoleControllerIntegrationTest.java [hunk 5]
```diff
@@ -48,4 +63,65 @@ void getJobRoleById_shouldReturn400_whenIDNotFound() {
...
+        Response response = APP.client().target("http://localhost:8080/api/job-roles")
+                .request()
+                .post(javax.ws.rs.client.Entity.json(jobRoleRequest));
+
+        Assertions.assertEquals(201, response.getStatus());
+    }
+
+    @Test
+    void createJobRole_shouldReturn400_whenInvalidData() {
+        jobRoleRequest.setName(null);
```
Comment (shaunganley) [inline src/test/java/integration/controller/JobRoleControllerIntegrationTest.java:78]: create the object inside the test method

File: src/test/java/integration/controller/JobRoleControllerIntegrationTest.java [hunk 6]
```diff
@@ -48,4 +63,65 @@ void getJobRoleById_shouldReturn400_whenIDNotFound() {
...
+                .post(javax.ws.rs.client.Entity.json(jobRoleRequest));
+
+        Assertions.assertEquals(400, response.getStatus());
+    }
+
+    @Test
+    void updateJobRole_shouldReturn200_whenUpdated() {
+        Response response = APP.client().target("http://localhost:8080/api/job-roles/2")
+                .request()
+                .put(javax.ws.rs.client.Entity.json(jobRoleRequest));
```
Comment (shaunganley) [inline src/test/java/integration/controller/JobRoleControllerIntegrationTest.java:90]: this test is dependent on role with id 2 existing in the db, instead you should create a role as part of this test,  get the new id and use it to update

File: src/test/java/integration/controller/JobRoleControllerIntegrationTest.java [hunk 7]
```diff
@@ -48,4 +63,65 @@ void getJobRoleById_shouldReturn400_whenIDNotFound() {
...
+                        1,
+                        "spec"
+                )));
+
+        Assertions.assertEquals(400, response.getStatus());
+    }
+
+    @Test
+    void deleteJobRole_shouldReturn200_whenDeleted() throws DatabaseConnectionException, InvalidJobRoleException, FailedToCreateJobRoleException {
+        int id = (new JobRoleService(new JobRoleDao(), new JobRoleValidator(new JobRoleDao(), new BandDao(), new CapabilityDao()))).createJobRole(jobRoleRequest);
```
Comment (shaunganley) [inline src/test/java/integration/controller/JobRoleControllerIntegrationTest.java:120]: you can just call the create a role endpoint as part of this test, get the new id and use it to delete

File: src/test/java/unit/service/JobRoleServiceTest.java [hunk 8]
```diff
@@ -75,4 +85,55 @@ void getJobRolesById_shouldThrowSQLException_whenDaoThrowsSQLException() throws
...
+    @Test
+    void createJobRole_shouldReturnId_whenDaoReturnsId() throws SQLException, DatabaseConnectionException, InvalidJobRoleException, FailedToCreateJobRoleException {
+        Mockito.when(jobRoleDao.createJobRole(jobRoleRequest)).thenReturn(1);
+        Mockito.when(jobRoleValidator.isValidJobRole(jobRoleRequest)).thenReturn(true);
+        assertEquals(1, jobRoleService.createJobRole(jobRoleRequest));
+    }
+
+    @Test
+    void updateJobRole_shouldReturnTrue_whenDaoReturnsTrue() throws SQLException, DatabaseConnectionException, InvalidJobRoleException, FailedToUpdateJobRoleException, JobRoleDoesNotExistException {
+        Mockito.when(jobRoleDao.createJobRole(jobRoleRequest)).thenReturn(1);
```
Comment (shaunganley) [inline src/test/java/unit/service/JobRoleServiceTest.java:98]: this shouldn't be mocked here

Comment (shaunganley) [reply src/test/java/unit/service/JobRoleServiceTest.java:98]: you're also missing the validator mock in here

File: src/test/java/unit/service/JobRoleServiceTest.java [hunk 9]
```diff
@@ -75,4 +85,55 @@ void getJobRolesById_shouldThrowSQLException_whenDaoThrowsSQLException() throws
...
+        Mockito.when(jobRoleDao.getJobRoleById(1)).thenReturn(jobRole);
+        Mockito.when(jobRoleDao.updateJobRole(1, jobRoleRequest)).thenReturn(true);
+        assertTrue(jobRoleService.updateJobRole(1, jobRoleRequest));
+    }
+
+    @Test
+    void updateJobRole_shouldThrowJobRoleDoesNotExistException_whenDaoReturnsNull() throws SQLException, DatabaseConnectionException, InvalidJobRoleException, FailedToUpdateJobRoleException {
+        Mockito.when(jobRoleDao.createJobRole(jobRoleRequest)).thenReturn(1);
```
Comment (shaunganley) [inline src/test/java/unit/service/JobRoleServiceTest.java:106]: this shouldn't be mocked here, you're also missing the validator mock in here

File: src/test/java/unit/service/JobRoleServiceTest.java [hunk 10]
```diff
@@ -75,4 +85,55 @@ void getJobRolesById_shouldThrowSQLException_whenDaoThrowsSQLException() throws
...
+        Mockito.when(jobRoleDao.getJobRoleById(1)).thenReturn(null);
+        assertThrows(JobRoleDoesNotExistException.class, () -> jobRoleService.updateJobRole(1, jobRoleRequest));
+    }
+
+    @Test
+    void updateJobRole_shouldThrowSQLException_whenDaoThrowsSQLException() throws SQLException, DatabaseConnectionException, InvalidJobRoleException {
+        Mockito.when(jobRoleDao.createJobRole(jobRoleRequest)).thenReturn(1);
```
Comment (shaunganley) [inline src/test/java/unit/service/JobRoleServiceTest.java:113]: you're also missing the validator mock in here

File: src/test/java/unit/service/JobRoleServiceTest.java [hunk 11]
```diff
@@ -75,4 +85,55 @@ void getJobRolesById_shouldThrowSQLException_whenDaoThrowsSQLException() throws
...
+    @Test
+    void deleteJobRole_shouldReturnTrue_whenDaoReturnsTrue() throws SQLException, DatabaseConnectionException, FailedToDeleteJobRoleException, JobRoleDoesNotExistException {
+        Mockito.when(jobRoleDao.getJobRoleById(1)).thenReturn(jobRole);
+        Mockito.when(jobRoleDao.deleteJobRole(1)).thenReturn(true);
+        assertTrue(jobRoleService.deleteJobRole(1));
+    }
+
+    @Test
+    void deleteJobRole_shouldThrowJobRoleDoesNotExistException_whenDaoReturnsFalse() throws SQLException, DatabaseConnectionException, FailedToDeleteJobRoleException {
+        Mockito.when(jobRoleDao.deleteJobRole(1)).thenReturn(false);
```
Comment (shaunganley) [inline src/test/java/unit/service/JobRoleServiceTest.java:135]: duplicate

File: src/test/java/unit/service/JobRoleServiceTest.java [hunk 12]
```diff
@@ -75,4 +85,55 @@ void getJobRolesById_shouldThrowSQLException_whenDaoThrowsSQLException() throws
...
+        Mockito.when(jobRoleDao.deleteJobRole(1)).thenReturn(false);
```
Comment (shaunganley) [inline src/test/java/unit/service/JobRoleServiceTest.java:136]: you're missing other mocks in this test

File: src/test/java/unit/service/JobRoleServiceTest.java [hunk 13]
```diff
@@ -75,4 +85,55 @@ void getJobRolesById_shouldThrowSQLException_whenDaoThrowsSQLException() throws
...
+        Mockito.when(jobRoleDao.createJobRole(jobRoleRequest)).thenReturn(1);
+        Mockito.when(jobRoleDao.getJobRoleById(1)).thenReturn(jobRole);
+        Mockito.when(jobRoleDao.updateJobRole(1, jobRoleRequest)).thenReturn(true);
+        assertTrue(jobRoleService.updateJobRole(1, jobRoleRequest));
+    }
+
+    @Test
+    void updateJobRole_shouldThrowJobRoleDoesNotExistException_whenDaoReturnsNull() throws SQLException, DatabaseConnectionException, InvalidJobRoleException, FailedToUpdateJobRoleException {
+        Mockito.when(jobRoleDao.createJobRole(jobRoleRequest)).thenReturn(1);
+        Mockito.when(jobRoleDao.getJobRoleById(1)).thenReturn(null);
```
Comment (shaunganley) [inline src/test/java/unit/service/JobRoleServiceTest.java:107]: need to test this scenario for the other endpoints

File: src/test/java/validator/JobRoleValidatorTest.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+            1,
+            "Engineering"
+    );
+
+
+    @Test
+    public void isValidJobRole_shouldReturnTrue_whenValidJobRole() throws InvalidJobRoleException, DatabaseConnectionException, SQLException {
+        Mockito.when(bandDao.getBandById(1)).thenReturn(band);
+        Mockito.when(capabilityDao.getCapabilityById(1)).thenReturn(capability);
+        assert(jobRoleValidator.isValidJobRole(jobRoleRequest));
```
Comment (shaunganley) [inline src/test/java/validator/JobRoleValidatorTest.java:52]: assertTrue

File: src/test/java/validator/JobRoleValidatorTest.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    public void isValidJobRole_shouldThrowInvalidJobRoleException_whenSpecificationTooLong() {
+        jobRoleRequest.setSpecification("X".repeat(256));
+        assertThrows(InvalidJobRoleException.class, () -> {
+            jobRoleValidator.isValidJobRole(jobRoleRequest);
+        });
+    }
+
+    @Test
+    public void isValidJobRole_shouldThrowInvalidJobRoleException_whenSpecificationTooShort() {
+        jobRoleRequest.setSpecification("");
```
Comment (shaunganley) [inline src/test/java/validator/JobRoleValidatorTest.java:113]: define the request object in the test method

File: src/test/java/validator/JobRoleValidatorTest.java [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+    public void isValidJobRole_shouldThrowInvalidJobRoleException_whenCapabilityIdIsNotFound() throws DatabaseConnectionException, SQLException {
+        Mockito.when(capabilityDao.getCapabilityById(1)).thenReturn(null);
+        assertThrows(InvalidJobRoleException.class, () -> {
+            jobRoleValidator.isValidJobRole(jobRoleRequest);
+        });
+    }
+
+    @Test
+    public void isValidJobRole_shouldThrowInvalidJobRoleException_whenSpecificationTooLong() {
+        jobRoleRequest.setSpecification("X".repeat(256));
```
Comment (shaunganley) [inline src/test/java/validator/JobRoleValidatorTest.java:105]: define the request object in the test method

File: src/test/java/validator/JobRoleValidatorTest.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+    public void isValidJobRole_shouldThrowInvalidJobRoleException_whenBandIdIsNotFound() throws DatabaseConnectionException, SQLException {
+        Mockito.when(bandDao.getBandById(1)).thenReturn(null);
+        assertThrows(InvalidJobRoleException.class, () -> {
+            jobRoleValidator.isValidJobRole(jobRoleRequest);
+        });
+    }
+
+    @Test
+    public void isValidJobRole_shouldThrowInvalidJobRoleException_whenCapabilityIdIsZeroOrLess() {
+        jobRoleRequest.setCapabilityId(0);
```
Comment (shaunganley) [inline src/test/java/validator/JobRoleValidatorTest.java:89]: define the request object in the test method

File: src/test/java/validator/JobRoleValidatorTest.java [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+    public void isValidJobRole_shouldThrowInvalidJobRoleException_whenNameTooShort() {
+        jobRoleRequest.setName("");
+        assertThrows(InvalidJobRoleException.class, () -> {
+            jobRoleValidator.isValidJobRole(jobRoleRequest);
+        });
+    }
+
+    @Test
+    public void isValidJobRole_shouldThrowInvalidJobRoleException_whenBandIdSsZeroOrLess() {
+        jobRoleRequest.setBandId(0);
```
Comment (shaunganley) [inline src/test/java/validator/JobRoleValidatorTest.java:73]: define the request object in the test method

File: src/test/java/validator/JobRoleValidatorTest.java [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+    public void isValidJobRole_shouldThrowInvalidJobRoleException_whenNameTooLong() {
+        jobRoleRequest.setName("X".repeat(51));
+        assertThrows(InvalidJobRoleException.class, () -> {
+            jobRoleValidator.isValidJobRole(jobRoleRequest);
+        });
+    }
+
+    @Test
+    public void isValidJobRole_shouldThrowInvalidJobRoleException_whenNameTooShort() {
+        jobRoleRequest.setName("");
```
Comment (shaunganley) [inline src/test/java/validator/JobRoleValidatorTest.java:65]: define the request object in the test method

File: src/test/java/validator/JobRoleValidatorTest.java [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+    @Test
+    public void isValidJobRole_shouldReturnTrue_whenValidJobRole() throws InvalidJobRoleException, DatabaseConnectionException, SQLException {
+        Mockito.when(bandDao.getBandById(1)).thenReturn(band);
+        Mockito.when(capabilityDao.getCapabilityById(1)).thenReturn(capability);
+        assert(jobRoleValidator.isValidJobRole(jobRoleRequest));
+    }
+
+    @Test
+    public void isValidJobRole_shouldThrowInvalidJobRoleException_whenNameTooLong() {
+        jobRoleRequest.setName("X".repeat(51));
```
Comment (shaunganley) [inline src/test/java/validator/JobRoleValidatorTest.java:57]: define the request object in the test method

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: src/main/java/org/kainos/ea/controller/JobRoleController.java (hunk 2 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/controller/JobRoleController.java:57]: done, think this happened after changing from ok to created and forgot to return the id again

File: src/main/java/org/kainos/ea/dao/BandDao.java (hunk 3 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/dao/BandDao.java:16]: done

File: src/main/java/org/kainos/ea/service/CapabilityService.java (hunk 4 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/service/CapabilityService.java:10]: sorted

File: src/test/java/integration/controller/JobRoleControllerIntegrationTest.java (hunk 5 above)
Comment (joshuasy10) [reply src/test/java/integration/controller/JobRoleControllerIntegrationTest.java:78]: i have added a @BeforeEach to reset the request object each time.

File: src/test/java/integration/controller/JobRoleControllerIntegrationTest.java (hunk 6 above)
Comment (joshuasy10) [reply src/test/java/integration/controller/JobRoleControllerIntegrationTest.java:90]: done

File: src/test/java/validator/JobRoleValidatorTest.java (hunk 16 above)
Comment (joshuasy10) [reply src/test/java/validator/JobRoleValidatorTest.java:105]: also added a @BeforeEach to reset the objects

File: src/test/java/validator/JobRoleValidatorTest.java (hunk 17 above)
Comment (joshuasy10) [reply src/test/java/validator/JobRoleValidatorTest.java:89]: see above

File: src/test/java/validator/JobRoleValidatorTest.java (hunk 18 above)
Comment (joshuasy10) [reply src/test/java/validator/JobRoleValidatorTest.java:73]: see above

File: src/test/java/validator/JobRoleValidatorTest.java (hunk 19 above)
Comment (joshuasy10) [reply src/test/java/validator/JobRoleValidatorTest.java:65]: see above

File: src/test/java/validator/JobRoleValidatorTest.java (hunk 20 above)
Comment (joshuasy10) [reply src/test/java/validator/JobRoleValidatorTest.java:57]: see above

File: src/main/java/org/kainos/ea/service/CapabilityService.java [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+    private final CapabilityDao capabilityDao;
+
+    public CapabilityService(CapabilityDao capabilityDao){
+        this.capabilityDao = capabilityDao;
+    }
+
+    public List<Capability> getAllCapabilities() throws FailedToGetCapabilityException, DatabaseConnectionException {
+        try {
+            return this.capabilityDao.getAllCapabilities();
+        } catch (SQLException e) {
```
Comment (ccurran01) [inline src/main/java/org/kainos/ea/service/CapabilityService.java:19]: Could this catch be extended to include the `DatabaseConnectionException`? That way both the 500 scenarios are covered by a single exception in the controller.

File: src/main/java/org/kainos/ea/service/JobRoleService.java [hunk 22]
```diff
@@ -37,4 +41,39 @@ public JobRole getJobRolesById(int id) throws FailedToGetJobRoleException, Datab
...
+    }
+
+    public boolean updateJobRole(int id, JobRoleRequest jobRoleRequest) throws FailedToUpdateJobRoleException, DatabaseConnectionException, InvalidJobRoleException, JobRoleDoesNotExistException {
+        try {
+            if(jobRoleDao.getJobRoleById(id) == null){
+                throw new JobRoleDoesNotExistException();
+            }
+            jobRoleValidator.isValidJobRole(jobRoleRequest);
+            return jobRoleDao.updateJobRole(id, jobRoleRequest);
+        } catch (SQLException e) {
```
Comment (ccurran01) [inline src/main/java/org/kainos/ea/service/JobRoleService.java:62]: Could this catch be extended to include the `DatabaseConnectionException`? That way both the 500 scenarios are covered by a single exception in the controller.

File: src/main/java/org/kainos/ea/service/JobRoleService.java [hunk 23]
```diff
@@ -37,4 +41,39 @@ public JobRole getJobRolesById(int id) throws FailedToGetJobRoleException, Datab
...
+    }
+
+    public boolean deleteJobRole(int id) throws FailedToDeleteJobRoleException, DatabaseConnectionException, JobRoleDoesNotExistException {
+        try {
+            if(jobRoleDao.getJobRoleById(id) == null){
+                throw new JobRoleDoesNotExistException();
+            }
+            return jobRoleDao.deleteJobRole(id);
+        } catch (SQLException e) {
+            System.err.println(e.getMessage());
```
Comment (ccurran01) [inline src/main/java/org/kainos/ea/service/JobRoleService.java:75]: Could this catch be extended to include the `DatabaseConnectionException`? That way both the 500 scenarios are covered by a single exception in the controller.

File: src/main/java/org/kainos/ea/validator/JobRoleValidator.java [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+        try {
+            if ( bandDao.getBandById(jobRoleRequest.getBandId()) == null) {
+                throw new InvalidJobRoleException("Band ID does not exist");
+            }
+            if ( capabilityDao.getCapabilityById(jobRoleRequest.getCapabilityId()) == null) {
+                throw new InvalidJobRoleException("Capability ID does not exist");
+            }
+            return true;
+        } catch (SQLException | DatabaseConnectionException e) {
+            throw new InvalidJobRoleException(e.getMessage());
```
Comment (ccurran01) [inline src/main/java/org/kainos/ea/validator/JobRoleValidator.java:48]: Should the details of an SQL or Database Connection error be exposed to the API consumers as a 400 error?

File: src/test/java/integration/controller/JobRoleControllerIntegrationTest.java [hunk 25]
```diff
@@ -48,4 +69,67 @@ void getJobRoleById_shouldReturn400_whenIDNotFound() {
...
```
Comment (ccurran01) [inline src/test/java/integration/controller/JobRoleControllerIntegrationTest.java:128]: Could this outstanding comment from @shaunganley be addressed. I think I agree with his suggestion of going through the endpoint.

File: src/test/java/unit/service/JobRoleServiceTest.java (hunk 8 above)
Comment (ccurran01) [reply src/test/java/unit/service/JobRoleServiceTest.java:98]: Agree with @shaunganley suggestions. Could we remove the create role mock, the job role create function isn't called during in the service method to update a role. The job role validator is called though and there is a missing when call for that

File: src/test/java/unit/service/JobRoleServiceTest.java (hunk 10 above)
Comment (ccurran01) [reply src/test/java/unit/service/JobRoleServiceTest.java:113]: Lets remove this when for the createJobRole and add in the when call for the validator

File: src/test/java/unit/service/JobRoleServiceTest.java (hunk 9 above)
Comment (ccurran01) [reply src/test/java/unit/service/JobRoleServiceTest.java:106]: Lets remove this when for the createJobRole and add in the when call for the validator

File: src/test/java/unit/service/JobRoleServiceTest.java (hunk 13 above)
Comment (ccurran01) [reply src/test/java/unit/service/JobRoleServiceTest.java:107]: This scenario seems to be missing for the delete and update functions.

File: src/test/java/unit/service/JobRoleServiceTest.java (hunk 11 above)
Comment (ccurran01) [reply src/test/java/unit/service/JobRoleServiceTest.java:135]: This when statement is duplicated with the one below

Comment (ccurran01) [reply src/test/java/unit/service/JobRoleServiceTest.java:135]: Are we not missing a when call for getJobRoleById

File: src/test/java/validator/JobRoleValidatorTest.java (hunk 14 above)
Comment (ccurran01) [reply src/test/java/validator/JobRoleValidatorTest.java:73]: Could we address @shaunganley's comment please

File: src/main/java/org/kainos/ea/service/BandService.java [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+    private BandDao bandDao;
+
+    public BandService(BandDao bandDao) {
+        this.bandDao = bandDao;
+    }
+
+    public List<Band> getAllBands() throws FailedToGetBandException, DatabaseConnectionException {
+        try {
+            return bandDao.getAllBands();
+        } catch (SQLException e) {
```
[the remark this answers is not in the document]
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/service/BandService.java:22]: done

File: src/main/java/org/kainos/ea/service/CapabilityService.java (hunk 21 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/service/CapabilityService.java:19]: done

File: src/main/java/org/kainos/ea/service/JobRoleService.java (hunk 22 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/service/JobRoleService.java:62]: done

File: src/main/java/org/kainos/ea/service/JobRoleService.java (hunk 23 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/service/JobRoleService.java:75]: done

File: src/main/java/org/kainos/ea/service/JobRoleService.java [hunk 27]
```diff
@@ -37,4 +41,39 @@ public JobRole getJobRolesById(int id) throws FailedToGetJobRoleException, Datab
             throw new FailedToGetJobRoleException();
         }
     }
+
+    public int createJobRole(JobRoleRequest jobRoleRequest) throws FailedToCreateJobRoleException, DatabaseConnectionException, InvalidJobRoleException {
+        try {
+            jobRoleValidator.isValidJobRole(jobRoleRequest);
+            return jobRoleDao.createJobRole(jobRoleRequest);
+        } catch (SQLException e) {
```
[the remark this answers is not in the document]
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/service/JobRoleService.java:49]: done

File: src/main/java/org/kainos/ea/validator/JobRoleValidator.java (hunk 24 above)
Comment (joshuasy10) [reply src/main/java/org/kainos/ea/validator/JobRoleValidator.java:48]: done

File: src/test/java/integration/controller/JobRoleControllerIntegrationTest.java (hunk 25 above)
Comment (joshuasy10) [reply src/test/java/integration/controller/JobRoleControllerIntegrationTest.java]: done

File: src/test/java/integration/controller/JobRoleControllerIntegrationTest.java [hunk 28]
```diff
@@ -48,4 +69,67 @@ void getJobRoleById_shouldReturn400_whenIDNotFound() {
...
+        Response response = APP.client().target("http://localhost:8080/api/job-roles/0")
+                .request()
+                .put(javax.ws.rs.client.Entity.json(jobRoleRequest));
+
+        Assertions.assertEquals(400, response.getStatus());
+    }
+
+    @Test
+    void updateJobRole_shouldReturn400_whenInvalidRequest() throws DatabaseConnectionException, InvalidJobRoleException, FailedToCreateJobRoleException {
+        int id = (new JobRoleService(new JobRoleDao(), new JobRoleValidator(new JobRoleDao(), new BandDao(), new CapabilityDao()))).createJobRole(jobRoleRequest);
```
[the remark this answers is not in the document]
Comment (joshuasy10) [reply src/test/java/integration/controller/JobRoleControllerIntegrationTest.java]: done

**Review by shaunganley — APPROVED** [review]

**Pull request merged**